carbon-design-system/carbon-addons-iot-react
# fix(composed-modal): adjust height based on content and available area

Closes #3077 

**Summary**

- This is a tricky issue, and I'm not certain I love my solution here, but wanted to gather some feedback. Using the refs on the modal elements, I'm measuring the footer/header to know their height to be able to subtract it from the whole modal and know the maxHeight of the content. This is set as a style object on the ModalBody; however, since the Carbon modal's css doesn't have a fixed height (it sets a min-height, and a max-height--and the max-height varies based on screen size)this approach doesn't completely solve the issue of the modal changing height as content grows. It will prevent it from growing too large, but the modal will still change size until it reaches that max-height set on the overall modal. Additionally, I'm setting a css var of the "extrasHeight" (header and footer), so that those can be read by components that use the ComposedModal to set their own height restrictions.

Old summary: ~I'm measuring the height of the window, the modal, the header and footer, and calculating the maximum height the content can be. If the modal is _not_ `isLarge` I set that height as the maxHeight and allow the modal to grow as needed; however, if it _is_ `isLarge` then, I set that as the `height` of the modal body, so that it maintains the same height all the time.~ 

**Change List (commits, features, bugs, etc)**

- measure the window and modal components to calculate a maximum height for the content
- set that height as the maxHeight on the modalBody style
- set a css var of the height of the header/footer combined.

**Acceptance Test (how to verify the PR)**

- check the SelectUserModal stories or TableColumnCustomizationModal stories
- open and close groups in SelectUserModal and see how modal stays the same height (it's using isLarge)
- check the "with many columns and 'Load more'" column customization story

**Regression Test (how to make sure this PR doesn't break old functionality)**

- confirm no height issues exist at various window sizes
- check the ComposedModal stories for regressions with isLarge knob being used or not.



**Things to look for during review**
